content/200-orm/200-prisma-client/150-using-raw-sql/100-typedsql.mdx

```diff
@@ -0,0 +1,204 @@
+
+   ```sql title="prisma/sql/getUsersByAge.sql"
+   SELECT id, name, age
+   FROM users
+   WHERE age > :minAge AND age < :maxAge
+   ```
+   </TabItem>
+   </TabbedContent>
+
+   :::note
+   
+   See below for information on how to type placeholders in your SQL files.
+
+   :::
+
+1. When using the generated function in your TypeScript code, pass the arguments as additional parameters to `$queryRawTyped`:
+
+   ```typescript
+   import { PrismaClient } from '@prisma/client'
+   import { getUsersByAge } from '@prisma/client/sql'
+
+   const prisma = new PrismaClient()
+
+   const minAge = 18
+   const maxAge = 30
+   const users = await prisma.$queryRawTyped(getUsersByAge(minAge, maxAge))
+   console.log(users)
+   ```
+
+By using parameterized queries, you ensure type safety and protect against SQL injection vulnerabilities. The TypedSQL generator will create the appropriate TypeScript types for the parameters based on your SQL query, providing full type checking for both the query results and the input parameters.
+
+### Passing array arguments to TypedSQL
+
+TypedSQL supports passing arrays as arguments for PostgreSQL. Use PostgreSQL's `ANY` operator with an array parameter.
+
+```sql title="prisma/sql/getUsersByIds.sql"
+SELECT id, name, email
+FROM users
+WHERE id = ANY($1)
+```
+
+```typescript
+import { PrismaClient } from '@prisma/client'
+import { getUsersByIds } from '@prisma/client/sql'
+
+const prisma = new PrismaClient()
+
+const userIds = [1, 2, 3]
+const users = await prisma.$queryRawTyped(getUsersByIds(userIds))
+console.log(users)
+```
+
+TypedSQL will generate the appropriate TypeScript types for the array parameter, ensuring type safety for both the input and the query results.
+
+:::note
+
+When passing array arguments, be mindful of the maximum number of placeholders your database supports in a single query. For very large arrays, you may need to split the query into multiple smaller queries.
+
+:::
+
+### Defining argument types in your SQL files
+
+Argument typing in TypedSQL is accomplished via specific comments in your SQL files. These comments are of the form:
+
+```sql
+-- @param {Type} $N:alias optional description
+```
+
+Where `Type` is a valid database type, `N` is the position of the argument in the query, and `alias` is an optional alias for the argument that is used in the TypeScript type.
+
+As an example, if you needed to type a single string argument with the alias `name` and the description "The name of the user", you would add the following comment to your SQL file:
+```sql
+-- @param {String} $1:name The name of the user
+```
+
+Currently accepted types are `Int`, `BigInt`, `Float`, `Boolean`, `String`, `DateTime`, `Json`, `Bytes`, and `Decimal`.
+
+Taking the [example from above](#passing-arguments-to-typedsql-queries), the SQL file would look like this:
+
+```sql
+-- @param {Int} $1:minAge
```

Comment:
We dont have a section about optional variables. We should add that. https://prisma-company.slack.com/archives/C058VM009HT/p1723564084984619